File: frontend/lib/core/models/vessel.dart
class Vessel {
  final String id;
  final String userId;
  final String name;
  final String make;
  final String model;
  final int? year;
  final String? hullMaterial;
  final double? lengthFt;
  final String? engineType;
  final String? engineMake;
  final String? engineModel;
  final String? imageUrl;
  final bool isPrimary;
  final DateTime createdAt;

  Vessel({
    required this.id,
    required this.userId,
    required this.name,
    required this.make,
    required this.model,
    this.year,
    this.hullMaterial,
    this.lengthFt,
    this.engineType,
    this.engineMake,
    this.engineModel,
    this.imageUrl,
    this.isPrimary = false,
    required this.createdAt,
  });

  factory Vessel.fromJson(Map<String, dynamic> json) {
    return Vessel(
      id: json['id'] as String,
      userId: json['user_id'] as String,
      name: json['name'] as String,
      make: json['make'] as String,
      model: json['model'] as String,
      year: json['year'] as int?,
      hullMaterial: json['hull_material'] as String?,
      lengthFt: (json['length_ft'] as num?)?.toDouble(),
      engineType: json['engine_type'] as String?,
      engineMake: json['engine_make'] as String?,
      engineModel: json['engine_model'] as String?,
      imageUrl: json['image_url'] as String?,
      isPrimary: json['is_primary'] as bool? ?? false,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  Map<String, dynamic> toJson() => {
        'name': name,
        'make': make,
        'model': model,
        if (year != null) 'year': year,
        if (hullMaterial != null) 'hull_material': hullMaterial,
        if (lengthFt != null) 'length_ft': lengthFt,
        if (engineType != null) 'engine_type': engineType,
        if (engineMake != null) 'engine_make': engineMake,
        if (engineModel != null) 'engine_model': engineModel,
        if (imageUrl != null) 'image_url': imageUrl,
        'is_primary': isPrimary,
      };

  Vessel copyWith({
    String? name,
    String? make,
    String? model,
    int? year,
    String? hullMaterial,
    double? lengthFt,
    String? engineType,
    String? engineMake,
    String? engineModel,
    String? imageUrl,
    bool? isPrimary,
  }) {
    return Vessel(
      id: id,
      userId: userId,
      name: name ?? this.name,
      make: make ?? this.make,
      model: model ?? this.model,
      year: year ?? this.year,
      hullMaterial: hullMaterial ?? this.hullMaterial,
      lengthFt: lengthFt ?? this.lengthFt,
      engineType: engineType ?? this.engineType,
      engineMake: engineMake ?? this.engineMake,
      engineModel: engineModel ?? this.engineModel,
      imageUrl: imageUrl ?? this.imageUrl,
      isPrimary: isPrimary ?? this.isPrimary,
      createdAt: createdAt,
    );
  }
}

File: backend/app/models/vessel.py
"""Vessel model — the My Vessel Garage."""

import uuid
from datetime import datetime

from sqlalchemy import String, Integer, Boolean, DateTime, ForeignKey
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class Vessel(Base):
    __tablename__ = "vessels"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    nickname: Mapped[str] = mapped_column(String, nullable=False)
    hin: Mapped[str | None] = mapped_column(String, nullable=True)
    make: Mapped[str] = mapped_column(String, nullable=False)
    model: Mapped[str] = mapped_column(String, nullable=False)
    year: Mapped[int] = mapped_column(Integer, nullable=False)
    hull_type: Mapped[str | None] = mapped_column(String, nullable=True)
    length_ft: Mapped[int | None] = mapped_column(Integer, nullable=True)
    primary_use: Mapped[str | None] = mapped_column(String, nullable=True)
    engine_make: Mapped[str | None] = mapped_column(String, nullable=True)
    engine_model: Mapped[str | None] = mapped_column(String, nullable=True)
    engine_serial: Mapped[str | None] = mapped_column(String, nullable=True)
    engine_hours: Mapped[int | None] = mapped_column(Integer, nullable=True)
    is_primary: Mapped[bool] = mapped_column(Boolean, default=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=datetime.utcnow
    )

    # Relationships
    user: Mapped["User"] = relationship("User", back_populates="vessels")  # noqa: F821
    ai_conversations: Mapped[list["AIConversation"]] = relationship(  # noqa: F821
        "AIConversation", back_populates="vessel"
    )
